### PR TITLE
build subfolders

### PR DESCRIPTION
**Pull request by dtothefp**
Here is my translation audit:
https://docs.google.com/a/optimizely.com/spreadsheets/d/1sZpLUxlU-pv9ncn0o5_NoDLysgjnZHCwC8TYXBhRpEg/edit#gid=0

Please put all translation issues into this doc, and check off below pages/layouts without issues
https://docs.google.com/a/optimizely.com/spreadsheets/d/1-n8_oWUWW6pwl0B_-McHuCDwJEIUSVUw7hZkkmThDVY/edit?usp=sharing

https://www.optimizelystaging.com/dfoxpowell/build-subfolders/de/
https://www.optimizelystaging.com/dfoxpowell/build-subfolders/fr/
https://www.optimizelystaging.com/dfoxpowell/build-subfolders/es/
https://www.optimizelystaging.com/dfoxpowell/build-subfolders/jp/

@4vanger test coverage is getting better. Run `grunt ui-test:assemble`

## Testing Plan
Please do not get hung up on layout issues cause by longer translated strings. This is pointless until the actual translated strings are in the site. Focus primarily if translations are missing. Also, important to note may be where `SL_{swap,hide}` behavior is missing and needs to be mimicked. 

Everyone:
- [x] *Layout*: WRAPPER
  - [x] /index (hompage)

@arthuracs 

- [x] *Layout*: TRIAL_FORM
  - [x] /free-trial
  - [x] /resources/ab-testing-tool
  - [x] /resources/split-testing-tool
- [x] *Layout*: FAQ
  - [x] /faq
  - [x] /gettingstarted
- [x] *Layout*: MODAL_WRAPPER
  - [x] /website-guts/templates/components/modals/** 
  - sign_up
  - sign_in
  - contact_sales
  - reset_password
  - live_demo_webinar (this is partially client side template)

@4vanger 
- [ ] *Layout*: ABOUT
  - [x] /about
  - [x] /contact
  - [x] /events 
  - [x] /press 
- [ ] *Layout*: CUSTOM
  - [x] /android
  - [x] /resources/live-demo-webinar
- [ ] *Layout*: BENEFITS
  - [x] /benefits
  - [x] /ecommerce -
  - [x] /enterprises
  - [x] /publishers
  - [x] /small-business

@stewsmith 
- [x] *Layout*: NO-SUB-NAV
  - [x] /partial/feature-list
  - [x] /feature-list
- [x] *Layout*: NO-SUB-NAV-OPTIONAL-HEADER
  - [x] /features-and-plans
  - [x] /partial/faq_section
  - [x] /partial/companies_panel
  - [x] /partial/features_list
  - [x] /pricing
- [x] *Layout*: SEO
  - [x] /ab-testing 
  - [x] /split-testing 
  - [x] /resources/conversion-optimization-tools
  - [x] /resources/ios-ab-testing
  - [x] /resources/multivariate-test-vs-ab-test
  - [x] /resources/multivariate-testing
  - [x] /resources/online-marketing
  - [x] /resources/website-optimization
- [x] *Layout*: RESOURCES
  - [x] /resources/index (check all individual resources render correctly and are translated)

@dtothefp
- [x] *Layout*: CUSTOMERS
  - [x] /customers/customer-stories
- [ ] *Layout*: PAGE
  - [x] /communities (removed target="blank" on mailto hrefs)
  - [x] /customers
  - [x] /demo
  - [x] /opt_out
  - [x] /partners/solutions
  - [x] /partners/technology
  - [x] /privacy
  - [x] /privacy/12/16/2013
  - [x] /security
  - [x] /resources/sample-size-calculator
  - [x] /terms-api
  - [x] /terms
- [x] *Layout*: PAGE_SUCCESS
  - [x] /success
- [x] *Layout*: STATISTICS
  - [x] /statistics
- [x] *Layout*: EVENTS_FORM
  - [x] /sxsw2015 **DO WE NEED THIS ANYMORE??**
- [ ] *Layout*: PARTNERS
  - [ ] / all solutions partners pages -- (individual partners description should not be translated)
  - [ ] / all technology partners pages -- (individual partners description should not be translated)
- [x] *Layout*: JOBS
  - [x] /jobs
- [x] *Layout*: MOBILE
  - [x] /mobile
- [x] *Layout*: NONPROFITS
  - [x] /nonprofits

#### TODO:
- @4vanger create client side `tr` helper
- @dtothefp omit pages from subfolders build (om, opticon)
- @dtothefp address existing `SL_{swap,hide}` using YML key replacement

### CODE REVIEW

File: grunt/assemble/plugins/smartling.js [hunk 1]
```diff
@@ -198,105 +163,14 @@ module.exports = function (assemble) {
...
-      } else {
-        console.log('Server catalog didn\'t change since last upload - using cached dictionaries.');
-        localeCodes.map(function(code){
-          var body = fs.readFileSync('tmp/download/' + code + '-' + DICT_FNAME, {encoding: 'UTF8'});
-          translations[code] = smartling.parsePOWithContext(body);
-        });
-        assemble.set('dicts', translations);
-        yamlDefer.resolve();
-      }
+      sendToSmartling(yamlDefer, jsDefer, phrases);
```
Comment (4vanger) [inline grunt/assemble/plugins/smartling.js]: It will be better if sendToSmartling will just return a single promise that you can use below. Or even convert it to callback

File: grunt/assemble/plugins/smartling.js [hunk 2]
```diff
@@ -198,105 +163,14 @@ module.exports = function (assemble) {
...
-      if(catalogChanged(content, 'tmp/upload/' + JS_DICT_FNAME)){
-        fs.writeFile('tmp/upload/' + JS_DICT_FNAME, content);
-        smartling.send(content, smartlingConfig.API_KEY, smartlingConfig.PROJECT_ID, JS_DICT_FNAME).then(function(){
-          var defers = localeCodes.map(function(code){
-            return smartling.fetch(smartlingConfig.URL, code, JS_DICT_FNAME, smartlingConfig.API_KEY, smartlingConfig.PROJECT_ID).then(function(body){
-              fs.writeFile('tmp/download/' + code + '-' + JS_DICT_FNAME, body);
-              deployJSDict(code, body);
-            });
-          });
+      Q.all([yamlDefer.promise, jsDefer.promise]).then(function(resolved){
```
Comment (4vanger) [inline grunt/assemble/plugins/smartling.js]: So it will look like
```
sendToSmartling(phrases).then(function(translations){
```
which is much more easy to read

File: grunt/assemble/plugins/smartling.js [hunk 3]
```diff
@@ -2,44 +2,19 @@
...
-    return true;
-  }
-  // extract list of keys (basically all the messages), sort them alphabetically and join together
-  var keys = Object.keys(smartling.parsePO(content)).sort().join('');
-  var latestKeys = Object.keys(smartling.parsePO(fs.readFileSync(fname, {encoding: 'UTF8'}))).sort().join('');
-  return keys !== latestKeys;
-}
+var generateKey = require('../utils/generate-key');
+var hbsParser = require('l10n-tools/hbs-parser');
+var objParser = require('l10n-tools/object-extractor');
```
Comment (4vanger) [inline grunt/assemble/plugins/smartling.js]: objParser is not used anymore

File: grunt/assemble/plugins/smartling.js [hunk 4]
```diff
@@ -2,44 +2,19 @@
...
+var Q = require('q');
 
 module.exports = function (assemble) {
+  var sendToSmartling = require('./translation-utils/smartling-upload')(assemble);
```
Comment (4vanger) [inline grunt/assemble/plugins/smartling.js]: If we will remove assemble dependency here - we will make smartling upload piece more independent from app and make it more testable and simple

Comment (4vanger) [inline grunt/assemble/plugins/smartling.js]: Took a look into smartling-upload.js - it does lots of tasks except actually uploading to smartling. If we will remove all of that stuff out there - we won't need assemble there.

File: grunt/assemble/plugins/smartling.js (hunk 3 above)
Comment (dtothefp) [inline grunt/assemble/plugins/smartling.js]: @4vanger can you please update this function with most recent syntax/libs for l10n

File: grunt/assemble/plugins/smartling.js (hunk 2 above)
Comment (dtothefp) [inline grunt/assemble/plugins/smartling.js]: @4vanger if you would like to refactor/update your code to be more testable it would be greatly appreciated because at this point I barely have time to refactor my own with all of the upcoming go to market projects in the pipeline for Q2

Comment (4vanger) [inline grunt/assemble/plugins/smartling.js]: this note is not related to my code - it is about the way how code was split into different files.

Comment (dtothefp) [inline grunt/assemble/plugins/smartling.js]: Then I recommend you break it into separate files in a way that you see most beneficial as I don't have knowledge of the inner workings of your code. I agree that returning a promise is a better idea but the point is I don't have time at this point to refactor and test both your code and mine.

File: grunt/assemble/plugins/smartling.js [hunk 5]
```diff
@@ -88,476 +59,129 @@ module.exports = function (assemble) {
...
-    //add the layout data
-    _.forEach(layoutData[websiteRoot], function(layoutObj, fp) {
+    var curryTryCatch = require('../utils/curry-try-catch');
+    var sendToSmartling = require('./translation-utils/smartling-upload')(assemble);
+    var populateSubfolderData = curryTryCatch(require('./translation-utils/populate-subfolder-data')(assemble));
+    var translatePageData = curryTryCatch(require('./translation-utils/translate-page-data')(assemble));
+    var mergeSubfolderYml = curryTryCatch(require('./translation-utils/merge-subfolder-yml')(assemble));
+    var mergeLayoutData = curryTryCatch(require('./translation-utils/merge-layout-data'));
+    var createTranslatedObject = curryTryCatch(require('./translation-utils/create-translated-object'));
+    var translateGlobalYml = curryTryCatch(require('./translation-utils/translate-global-yml'));
```
Comment (dtothefp) [inline grunt/assemble/plugins/smartling.js]: @4vanger separated all plugin functions into separate files and wrap in `curryTryCatch` closure to emit function error messages with function name to the assemble task

File: grunt/assemble/test/yml-file-data-spec.js [hunk 6]
```diff
@@ -0,0 +1,44 @@
+ var expect = require('chai').expect;
+ var assemble = require('assemble');
+ var readYml = require('../transforms/translation-helpers/yml-file-data');
+ var path = require('path');
+
+ describe('transforms', function() {
```
Comment (dtothefp) [inline grunt/assemble/test/yml-file-data-spec.js]: @4vanger add initial tests

File: grunt/assemble/utils/curry-try-catch.js [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+  var name = fn.name;
+
+  return function(){
+    try {
+      return fn.apply(fn, arguments);
+    } catch(error) {
+      throw new Error('Function Name: ' + name +  error);
+    }
+  };
+};
```
Comment (dtothefp) [inline grunt/assemble/utils/curry-try-catch.js]: @4vanger `curryTryCatch` function for logging errors in `require`d functions

**Comment by dtothefp** [comment]
retest this please

**Comment by stewsmith** [comment]
@dtothefp We don't need /events/sxsw2015 anymore

**Comment by dtothefp** [comment]
retest this please
